lib/moab/storage_object_validator.rb

```diff
@@ -0,0 +1,196 @@
+        version_sub_dirs = sub_dirs(version_path)
+        # Need to add before_result_size and after_result_size because if everything is expected
+        # and good, then no error message will be added to the result array, so if it stays the same
+        # size then it is ready to be checked for the data_sub_dirs and manfist_xml
+        before_result_size = results.size
+        results.concat check_sub_dirs(version_sub_dirs, version_dir, EXPECTED_VERSION_SUB_DIRS)
+        after_result_size = results.size
+        # don't bother checking the data sub dir if we didn't find the expected two subdirs
+
+        # Before it was result.size === 0, however for each version that is not correctly constructed,
+        # error messages will be concat to the array. By the time the loop reaches a version that is
+        # correctly constructe the size will stay the same as before (because it has the old error messages)
+        # so we must make sure before and after result size are the same - aka no error message was added
+        # and the version directory was correctly constructed
+
+        if before_result_size == after_result_size
+          data_dir_path = "#{version_path}/#{version_sub_dirs[0]}"
+          data_sub_dirs = sub_dirs(data_dir_path)
+          results.concat check_sub_dirs(data_sub_dirs, version_dir, EXPECTED_DATA_SUB_DIRS)
+          results.concat check_manifest_files(version_path, version_dir)
+        end
+      end
+
+      results
+    end
+
+    def check_sub_dirs(sub_dirs, version, required_sub_dirs)
+      results = []
+      sub_dir_count = sub_dirs.size
+      if sub_dir_count == 2
+        results.concat expected_dirs(sub_dirs, version, required_sub_dirs)
+      elsif sub_dir_count > 2
+        results.concat found_unexpected(sub_dirs, version, required_sub_dirs)
+      elsif sub_dir_count < 2
+        results.concat missing_data(sub_dirs, version, required_sub_dirs)
+      elsif sub_dir_count.zero?
+        results << result_hash(EMPTY)
+      end
+      results
+    end
+
+    def directory_entries(path)
+      @directory_entries_hash[path] ||=
+        begin
+          ret_val = { files: [], dirs: [] }
+          (Dir.entries(path).sort - IMPLICIT_DIRS).each do |child|
+            if FileTest.file? child
+              ret_val[:files] << child
+            else
+              ret_val[:dirs] << child
+            end
+          end
+          ret_val
+        end
+    end
+
+    def sub_dirs(path)
+      directory_entries(path)[:dirs]
+    end
+
+    def files_in_dir(path)
+      directory_entries(path)[:files]
+    end
+
+    def found_unexpected(array, version, required_sub_dirs)
+      results = []
+      unexpected = (array - required_sub_dirs) # removed .pop for now
+      unexpected = "#{unexpected} Version: #{version}"
+      results << result_hash(EXTRA_CHILD_DETECTED, unexpected)
+      results
+    end
+
+    def missing_data(array, version, required_sub_dirs)
```

Comment:
if this is about MISSING_DIR, should the method name be missing_dir ??  (error code and method name aren't matching.